Bran - Ladder step1

june20516 opened the pull request:
Step1 완성했습니다. 피드백 부탁드립니다.

june20516 committed:
사다리게임 - step1

값을 받는 함수
값을 확인, UInt로 변환하는 함수
세로 줄 만드는 함수
가로줄 만드는 함수
가로줄 만들 때 랜덤요소 결정하는 함수
홀짝줄 판단하는 함수
전체 사다리(2차원 배열) 만드는 함수
까지 완료

june20516 committed:
사다리게임 - step1

완성된 모양의 String로 만들어 출력하는 함수 작성.

june20516 committed:
사다리게임 - step1

출력부분 함수 분리
참가자, 게임결과 부분 생성
메인 함수를 만들어 함수간 흐름 정리
함수별 기능 주석으로 작성

godrm commented:
commit 하는 단위를 더 작은 변화로 나눠서 커밋해보세요. 
사다리게임 - step1 같은걸 작성할 필요없습니다. 저장소 이름과 브랜치 이름으로 구분이 가능하니까요.
좋은 커밋 로그를 작성하는 방법에 대해서도 공부해서 정리해보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+}
+
+//가로 줄 랜덤 결정
+func SetBarOrNil() -> String {
```
함수 이름이 대문자로 시작하면 안됩니다. 좀더 자연스러운 함수 이름을 작성해보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+    if let columnString = optionalColumnString, let rowString = optionalRowString {
+        return (columnString, rowString)
+    }
+    return ("nil", "nil")
```
receiveInput() 함수는 입력도 두 가지 동작을 하고 있네요.
사람수 입력 받는 함수와 사다리 높이 받는 함수를 분리해보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+func preventWrongInput(checkedInput: (columnNumber: UInt,rowNumber: UInt)) -> Void {
+    if checkedInput.columnNumber == 0 || checkedInput.rowNumber == 0 {
+        print("입력값이 잘못되었습니다.(1이상의 정수만 가능)")
+    }
```
예외처리를 추가한 것은 좋네요. 근데 1줄 짜리 혹은 1사람 사다리 게임이 동작가능한가요?

june20516 replied:
말씀해 주신 사항들 반영해서 수정했습니다. 확인 부탁드립니다! 커밋 단위를 쪼개는 것과 로그 작성하는 방법도 점차 개선해 나가겠습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+    while i < columnNumber*2-1 {
+        participants[i] = String(j)
+        j += 1
+        i += 2
```
i, j 같은 변수명은 의도를 알수 없는 변수입니다. 용도에 맞는 변수명을 사용해보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+    let checkedInput = checkInputValue(columnString: inputValue.columnString, rowString: inputValue.rowString)
+    preventWrongInput(checkedInput: checkedInput)
+    let wholeLadderKit = createWholeLadderKit(columnNumber: checkedInput.columnNumber, rowNumber: checkedInput.rowNumber)
+    let completedLadder = assembleWholeLadder(wholeLadderKit: wholeLadderKit)
```
`assembleWholeLadder(wholeLadderKit:)` 같은 동작을 하는 함수를 만들기 보다는
wholeLadderKit 을 받아서 원하는 형태로 출력하는 함수를 만드세요. 
출력을 위해서 긴 문자열을 만드는 것은 좋은 선택이 아닙니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+func main(inputValue: (columnString: String,rowString: String)) -> Void {
+    let checkedInput = checkInputValue(columnString: inputValue.columnString, rowString: inputValue.rowString)
+    preventWrongInput(checkedInput: checkedInput)
+    let wholeLadderKit = createWholeLadderKit(columnNumber: checkedInput.columnNumber, rowNumber: checkedInput.rowNumber)
```
wholeLadderKit 가 2차원 배열이라는 느낌이 좀 부족하네요. 복수로 표현할 수 있는 단어가 좋습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+//메인함수
+func main(inputValue: (columnString: String,rowString: String)) -> Void {
+    let checkedInput = checkInputValue(columnString: inputValue.columnString, rowString: inputValue.rowString)
+    preventWrongInput(checkedInput: checkedInput)
```
preventWrongInput()는 입력값으로 경고만 하는건가요?
예외처리를 하면 흐름제어가 들어가야 합니다. return을 하던 break를 하던 다시 입력받던 흐름제어를 만들어보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+    let wholeLadderKit = createWholeLadderKit(columnNumber: checkedInput.columnNumber, rowNumber: checkedInput.rowNumber)
+    let completedLadder = assembleWholeLadder(wholeLadderKit: wholeLadderKit)
+    let participants = makeParticipant(columnNumber: Int(checkedInput.columnNumber))
+    let gameResults = makeLosingTicket(columnNumber: Int(checkedInput.columnNumber))
```
원래 요구사항에 없던 기능을 구현한 거죠?
단지 출력을 위한 데이터를 만드는 거라면 출력하는 하위 모듈 내부에 넣는게 좋을 것 같습니다.
이 부분이 main에 있을만큼 중요한 기능이 아닌 것 같습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,136 @@
+//전체 배열 조립
+func createWholeLadderKit(columnNumber: UInt, rowNumber: UInt) -> [[String]] {
+    var ladder: [[String]] = Array()
+    for i in 0..<columnNumber*2-1 {
```
왜 데이터를 입력받은 columnNumber의 두 배만큼 사용하는건가요?
2배만큼 데이터를 사용해야만 하는 이유가 있을까요?

godrm reviewed (commented):
변수이름, 함수이름을 선택할 때 좀 더 주의해주세요.
다음 피드백을 확인해보세요.

june20516 committed:
입력 값의 종류에 따라 함수 분리
> 사다리 구조를 배열당 세로줄에서 배열당 가로층계로 변경

june20516 committed:
피드백 반영했습니다.
기존 로직은 부분배열에 세로열을 담았고, 때문에 크기를 항상 인원수*2-1 해야 했습니다. 가로방향으로 중간사다리만 데이터화하고, 나머지는 출력부에서 구현했습니다.
예외처리 함수가 결과에 따라 메인함수를 종료시키게 했습니다.
참가번호와 꽝도 출력부에서 처리하게 했습니다.
전체적인 변수명을 수정했습니다.

godrm commented:
마지막 커밋처럼 `피드백 반영했습니다.. + 부연 설명` 은 코멘트로 남기세요. 
커밋 로그에는 항상 소스 코드에 대한 작업 내용만 남기세요. 커밋은 저만 리뷰하라고 남기는게 아니라 본인이 작업한 걸 기록하기 위한 겁니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -7,17 +7,23 @@
+    if let columnString = readLine() {
+    return columnString
+    }
+    return "nil"
```
scope 내부에서는 항상 들여쓰기를 주의하세요. 
optional 을 넘기는 것과 "nil" 값을 넘기는 게 어떤 차이가 있을까요?
값에 특정한 예외값을 지정하는 것은 좋지 않습니다. 차라리 optional 을 처리하는게 더 낫습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -29,20 +35,17 @@ func checkInputValue(columnString: String, rowString: String) -> (columnNumber:
+//----------------------------------------------------------------------------
 //가로 줄 랜덤 결정
-func SetBarOrNil() -> String {
+func setLadderStairs() -> String {
```
이전보다 더 좋아졌습니다. `Stairs`가 복수면 String 내부에 여러 사다리 정보가 표현되나요?

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -51,63 +54,66 @@ func SetBarOrNil() -> String {
+//    } else {
+//        return createHorizontalLines(rowNumber: rowNumber)
+//    }
+//}
```
사용하지 않는 함수라면 과감하게 지우세요. 커밋 이력을 보면 이전 함수가 남아있으니까요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -129,6 +137,8 @@ func printGame(completedLadder: String, makeParticipant: [String], makeLosingTik
 
 
+//-----------------------------------------------------------------------------
+
```
함수 사이에 줄바꿈 주석을 넣는 경우도 있고 안 넣는 경우도 있고 어떤 기준인거죠?
주석이 코드 읽는 흐름을 방해할 정도만 아니면 상관없지만, 코드에 전혀 쓸모없는 주석은 의미가 없습니다. 주석도 의도가 있을때만 정확하게 하는 습관이 필요합니다. 참고하세요.

june20516 replied:
피드백 모두 반영 완료했습니다. 

역할별 부분으로 나누었는데 의미 전달이 되도록 재배치하고 주석에 내용을 넣었습니다.
입력받는 옵셔널 값은 그대로 넘겨 값을 확인하는 checkNumericInput 함수에서 처리하도록 변경했습니다.
참가자, 꽝 배열도 길이에 맞게 생성되고, 출력할 때 간격을 조절할 수 있도록 바꿨습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -51,63 +54,66 @@ func SetBarOrNil() -> String {
+    while index < columnNumber*2-1 {
+        participants[index] = String(personNumber)
+        personNumber += 1
+        index += 2
```
여전히 participants 는 2배 크기 배열로 만드는게 그대로군요. 
participants를 columnNumber 만큼만 만드는 방법은 없을까요?

june20516 replied:
출력부분에서 separator를 " "으로 두어 해결했습니다. 감사합니다 !

godrm reviewed (commented):
아래 피드백 내용 한 번만 더 확인해보고 넘어갑시다.

june20516 committed:
참가자 번호 생성 방식 개선

june20516 committed:
꽝 생성 방식 개선

june20516 committed:
입력값의 옵셔널 처리방식 변경

optional처리를 기존 receive 함수에서 check함수로 이동

june20516 committed:
사다리 세로 요소 구현 방식 변경

기존 배열의 값에 포함했던 것에서 출력시 separator로 출력되도록 변경
구분선 각주 역할별로 수정

june20516 committed:
checkNumericInput 코드 리팩토링

nil-coalesce 표현을 사용하여 간결화

godrm commented:
다음 단계를 진행하면서 함수가 가장 작은 단위로 하나의 역할만 담당하도록 개선해보세요.
